org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/commands/SemanticTokensCommand.java

```diff
@@ -38,22 +34,17 @@ public static SemanticTokens provide(String uri) {
 	}
 
 	private static SemanticTokens doProvide(String uri) {
-		IDocument document = null;
-
 		ITypeRoot typeRoot = JDTUtils.resolveTypeRoot(uri);
-		if (typeRoot != null) {
-			try {
-				document = JsonRpcHelpers.toDocument(typeRoot.getBuffer());
-			} catch (JavaModelException e) {
-				JavaLanguageServerPlugin.logException("Failed to provide semantic tokens for " + uri, e);
-			}
-		}
-		if (document == null) {
+		if (typeRoot == null) {
 			return new SemanticTokens(Collections.emptyList());
 		}
 
-		SemanticTokensVisitor collector = new SemanticTokensVisitor(document, SemanticTokenManager.getInstance());
 		CompilationUnit root = CoreASTProvider.getInstance().getAST(typeRoot, CoreASTProvider.WAIT_YES, new NullProgressMonitor());
+		if (root == null) {
+			return new SemanticTokens(Collections.emptyList());
+		}
+
+		SemanticTokensVisitor collector = new SemanticTokensVisitor(root, SemanticTokenManager.getInstance());
```

Comment:
Looks good. But not sure whether it can fix the problem completely.

As my understanding, both document and CompilationUnit are a snapshot of the document at some point. The difference is CompilationUnit is the AST tree parsed from the document at some point. I'm curious how the fix can guarantee it's not out-of-sync.

Reply:
Yes, the compilation unit might get out-of-sync with the real document at some points. But the problem before this change was that the semantic token visitor used both a document **and** AST nodes from the compilation unit, and the document was not in sync with the compilation unit.

Do you think we should also check for document updates during the request and recalculate the tokens? From my testing, the issue where semantic tokens get offset due to frequent updates to the document seems to be fixed by this PR. But please do verify this if you know how to reproduce that issue.

Reply:
Actually I just submitted redhat-developer/vscode-java#1632, which should fix the problem that you've pointed out. It turns out that there were two separate issues causing the offset, [see my comment](https://github.com/redhat-developer/vscode-java/issues/1597#issuecomment-696648993).